processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeDeriver.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.apm.processor.zipkin;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.hawkular.apm.api.model.Property;
+import org.hawkular.apm.api.model.events.CompletionTime;
+import org.hawkular.apm.server.api.model.zipkin.Annotation;
+import org.hawkular.apm.server.api.model.zipkin.Span;
+import org.hawkular.apm.server.api.services.SpanCache;
+import org.hawkular.apm.server.api.task.AbstractProcessor;
+import org.hawkular.apm.server.api.task.RetryAttemptException;
+import org.jboss.logging.Logger;
+
+/**
+ * Deriver for completion time of trace reported by zipkin instrumentation.
+ *
+ * @author Pavol Loffay
+ */
+@Singleton
+public class CompletionTimeDeriver extends AbstractProcessor<CompletionTimeProcessing, CompletionTime>{
+
+    private static final Logger log = Logger.getLogger(CompletionTimeDeriver.class.getName());
+
+    private SpanCache spanCache;
+
+
+    public CompletionTimeDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTime processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        /**
+         * TODO HWKAMP-348
+         *
+         * Wait one retry delay before calculating trace completion time. This is not a perfect solution but
+         * there is no information in parent span whether there are any children. Solution might be calculating
+         * trace completion time for all spans (not only for root as it is now) and update duration values directly
+         * in database.
+         */
```

Comment:
Updating database multiple times is not ideal, as depending on order of processing async events, the last update may not be the latest time. Other issue is that this processor is producing a CompletionTime event for consumption by other components - so if we produce multiple CompletionTime events for same trace instance, then that would need to be defined so that consumers are aware - but again suffers from issue of not knowing, at the time the span is processed, whether it is the latest (in terms of duration).